qwerty_mlir/lib/Qwerty/Transforms/QwertyToQCircConversionPass.cpp

```diff
@@ -2986,6 +2877,875 @@ void shootdownStandardizations(llvm::SmallVectorImpl<Standardization> &left_stdi
+      qwerty::QBundleBasisTranslationOp std_to_bar =
+          rewriter.create<qwerty::QBundleBasisTranslationOp>(
+              loc, std_1, bar_1, bar_phases, packed_last_qubit);
+
+      // 4. qbpack resulting qubits with first N-1 qubits
+      llvm::SmallVector<mlir::Value> final_qubits(
+          fwd_unpacked.begin(), std::prev(fwd_unpacked.end()));
+
+      mlir::ValueRange bar_qubit =
+          rewriter
+              .create<qwerty::QBundleUnpackOp>(loc, std_to_bar.getQbundleOut())
+              .getQubits();
+      final_qubits.push_back(bar_qubit.front());
+
+      // 5. replace op
+      rewriter.replaceOpWithNewOp<qwerty::QBundlePackOp>(trans, final_qubits);
+    } else { // foo // bar.revolve >> std**N
+      // id**(N-1) * (bar >> std) | foo // std.revolve >> std**N (reverse case)
+      // 1. unpack input bundle to get individual qubits
+      mlir::ValueRange in_unpacked =
+          rewriter.create<qwerty::QBundleUnpackOp>(loc, trans.getQbundleIn())
+              .getQubits();
+
+      // 2. take the last qubit, apply inverse std >> bar translation
+      llvm::SmallVector<mlir::Value> last_qubit(std::prev(in_unpacked.end()),
+                                                in_unpacked.end());
+      mlir::Value packed_last_qubit =
+          rewriter.create<qwerty::QBundlePackOp>(loc, last_qubit);
+
+      // create inverse std >> bar
+      qwerty::QBundleBasisTranslationOp bar_to_std =
+          rewriter.create<qwerty::QBundleBasisTranslationOp>(
+              loc, bar_1, std_1, bar_phases, packed_last_qubit);
+
+      // 3. unpack the result
+      mlir::ValueRange bar_std_unpacked =
+          rewriter
+              .create<qwerty::QBundleUnpackOp>(loc, bar_to_std.getQbundleOut())
+              .getQubits();
+
+      // 4. combine with first N-1 qubits
+      llvm::SmallVector<mlir::Value> combined_qubits(
+          in_unpacked.begin(), std::prev(in_unpacked.end()));
+
+      combined_qubits.push_back(bar_std_unpacked.front());
+
+      auto combined_packed =
+          rewriter.create<qwerty::QBundlePackOp>(loc, combined_qubits);
+
+      // 5. apply foo // std.revolve >> std**N
+      qwerty::QBundleBasisTranslationOp recurse_case_rev =
+          rewriter.create<qwerty::QBundleBasisTranslationOp>(
+              loc, revolve_basis, std_N, foo_phases, combined_packed);
+
+      mlir::ValueRange recurse_case_unpacked =
+          rewriter
+              .create<qwerty::QBundleUnpackOp>(loc,
+                                               recurse_case_rev.getQbundleOut())
+              .getQubits();
+
+      llvm::SmallVector<mlir::Value> final_qubits(recurse_case_unpacked.begin(),
+                                                  recurse_case_unpacked.end());
+
+      rewriter.replaceOpWithNewOp<qwerty::QBundlePackOp>(trans, final_qubits);
+    }
+
+    return mlir::success();
+  }
+};
+
+// TODO: Figure out how to do predicated RevolveGenerators like
+// (aka '1' * std[N] >> '1' * (std[N-1] // std.revolve))
```

Comment:
What do you think about making a GitHub issue for this? (We can leave the comment here too, why not)